src/back/msl/writer.rs

```diff
@@ -594,8 +608,28 @@ impl<W: Write> Writer<W> {
         log::trace!("expression {:?} = {:?}", expr_handle, expression);
         match *expression {
             crate::Expression::Access { base, index } => {
+                let accessing_wrapped_array =
+                    match *context.info[base].ty.inner_with(&context.module.types) {
+                        crate::TypeInner::Array { .. } => true,
+                        crate::TypeInner::Pointer {
+                            base: pointer_base,
+                            class,
+                        } => match context.module.types[pointer_base].inner {
+                            crate::TypeInner::Array { .. } => {
+                                class == crate::StorageClass::Function
```

Comment:
why are we limiting this only to Function and Private classes?

Reply:
I wasn't sure what other classes count. I don't think we want to be doing `.inner` on storage or uniform arrays though.

Reply:
Supposing I have a struct
```rust
struct Foo {
  array: array<f32; 3>;
};
```
So if this structure is a uniform/storage buffer, then we don't have `inner`.
But if it's a copy we made into function/private storage, it does have inner? It seems to me that we'd not even be able to re-use this structure when copying it, would we?

Reply:
I'm not entirely sure what you mean. If you could come up with a test shader for this then I would be grateful.
Something like:
```glsl
struct Foo {
  float[3] array;
};

Foo foo() {
    Foo foo = {{1.0, 1.5, 2.5}};
    return foo;
}
```
works fine though:
```metal
struct type2 {
    float inner[3u];
};
struct Foo {
    type2 array;
};
constant type2 const_type2_ = {1.0, 1.5, 2.5};
constant Foo const_Foo = {const_type2_};

Foo foo(
) {
    Foo foo1;
    foo1 = const_Foo;
    return foo1;
}

void main1(
    thread metal::float4& colour
) {
    Foo foo2;
    Foo _e14 = foo();
    foo2 = _e14;
    colour = float4(foo2.array.inner[2]);
    return;
}
```

Reply:
In your case, what happens if `Foo` is used as uniform/storage buffer? If you have a pointer to it, it will not be in Private/Function storage class, so the code in question will not trigger.

Reply:
Ah, good point. I've fixed this now.